File: marg/src/marg/token/TokenSet.java
package marg.token;


import marg.exception.ParseException;

import java.io.EOFException;
import java.util.List;

import static marg.token.TokenKind.*;

public class TokenSet {
	private final List<Token> list;
	private final int length;
	private int offset = 0;
	
    public TokenSet(List<Token> tokens) {
        list = tokens;
        length = tokens.size();
    }
	
	public void unget() {
		if (offset > 0) offset--;
	}
	public void unget(int k) {
		if (offset >= k) offset -= k;
	}
	public Token next() throws EOFException {
        checkEOF();
        return list.get(offset++);
	}
	public Token get() {
		return list.get(offset);
	}

    public boolean is(String s) {
        return s.equals(get().string);
    }

	public boolean isEOF() {
        return get().isEOF();
	}
    public void checkEOF() throws EOFException {
        if (isEOF()) throw new EOFException();
    }
	
    public boolean read(String... args) throws EOFException {
	    for (String t : args) {
	 		Token next = next();
	 		if (!t.equals(next.string)) return false;
		}
		return true;
	}

    public void readP(String... args) throws ParseException, EOFException {
        for (String t : args) {
            if (isEOF()) throw new EOFException();
            Token next = next();
            if (!t.equals(next.string))
                throw new ParseException();
        }
    }
	
	public boolean isName() {
		return isEOF()? false : get().kind == Identifier;
	}

	public boolean isNumber() {
		return isEOF()? false : get().kind == IntLiteral;
	}

    public boolean isBool() {
        return isEOF()? false : get().kind == BoolLiteral;
    }

	public boolean isOperator() {
		return isEOF()? false : get().kind == Operator;
	}

	public boolean isReserved() {
		return isEOF()? false : get().kind == Reserved;
	}
}

File: marg/src/marg/token/TokenSet.scala
package marg.token

import java.io.EOFException

import marg.exception.ParseException
import marg.token.TokenKind._

class TokenSet(ls: List[Token]) {

  def getOffset = offset

  def unget(): Unit =
    if (offset > 0) offset -= 1

  def unget(k: Int): Unit =
    if (offset >= k) offset -= k

  def next: Token = {
    checkEOF()
    offset += 1
    ls(offset - 1)
  }

  def get: Token = ls(offset)

  def is(s: String): Boolean = { s == get.String }

  def isEOF: Boolean = get.isEOF

  def checkEOF() = if (isEOF) throw new EOFException

  def read(args: String*): Boolean = {
    for (t <- args) {
      val n: Token = next
      if (t != n.String) return false
    }
    true
  }

  def readP(message: String, args: String*) {
    args.foreach(t => {
      checkEOF()
      if (next.String != t)
        throw new ParseException(s"Syntax Error: $message", this)
    })
  }


  def isEOL = !isEOF && (get.Kind eq EOL)

  def isName = !isEOF && (get.Kind eq Identifier)

  def isNumber = !isEOF && (get.Kind eq IntLiteral)

  def isBool = !isEOF && (get.Kind eq BoolLiteral)

  def isOperator = !isEOF && (get.Kind eq Operator)

  def isReserved = !isEOF && (get.Kind eq Reserved)

  private var offset = 0
}
